apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.Data/ClientPool.cs

```diff
@@ -0,0 +1,100 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using Google.Api.Gax.Grpc;
+using Google.Apis.Auth.OAuth2;
+using Google.Cloud.Spanner.V1;
+using Google.Cloud.Spanner.V1.Logging;
+using Grpc.Auth;
+
+namespace Google.Cloud.Spanner
+{
+    internal static class ClientPool
+    {
+        private static readonly ConcurrentDictionary<ClientPoolKey, ClientPoolEntry> s_clientEntryPool =
+            new ConcurrentDictionary<ClientPoolKey, ClientPoolEntry>();
+
+        public static int Timeout { get; set; }
+
+        public static Task<SpannerClient> AcquireClientAsync()
+        {
+            return AcquireClientAsync(null, null);
+        }
+
+        public static async Task<SpannerClient> AcquireClientAsync(ITokenAccess credentials, ServiceEndpoint endpoint)
+        {
+            var key = new ClientPoolKey {
+                Credential = credentials ?? ApplicationDefault.Instance,
+                Endpoint = endpoint ?? SpannerClient.DefaultEndpoint
+            };
+            var poolEntry = s_clientEntryPool.GetOrAdd(key, k => new ClientPoolEntry(key));
+            var result = await poolEntry.AcquireClientFromEntryAsync().ConfigureAwait(false);
+            Logger.LogPerformanceCounter("SpannerClient.Count", () => s_clientEntryPool.Count);
+            return result;
+        }
+
+        /// <summary>
+        ///     Callers should be careful not to call AcquireClient while CloseAll is still running.
```

Comment:
Nit: typically XML documentation isn't indented like this. It won't affect the output, but if you can tweak your plugin settings to avoid it, it would look more like the rest of the code.